pkg/csi/service/wcp/controller.go

```diff
@@ -555,9 +564,36 @@ func (c *controller) createBlockVolume(ctx context.Context, req *csi.CreateVolum
 					TopologyRequirement: topologyRequirement,
 					Vc:                  vc})
 			if err != nil {
+				// If the error is of InvalidTopologyProvisioningError type, it means we cannot
+				// recover from this error with a retry, so cleanup the voume created above.
+				if _, ok := err.(*common.InvalidTopologyProvisioningError); ok {
+					log.Errorf("Encountered error after creating volume. Cleaning up...")
+					// Delete the CnsVolumeOperationRequest created for CreateVolume call above.
+					deleteOpReqError := OperationStore.DeleteRequestDetails(ctx, req.Name)
```

Comment:
Can we do this clean-up after the volume is deleted?

Reply:
CRD should be removed only after the volume is deleted successfully, so we do not end up creating orphan volumes.

Reply:
I am doing the opposite actually. If I delete the volume but the CRD deletion fails for some reason, we will be stuck on a failure loop in idempotency forever. Here, I am deleting the CRD first and then deleting the volume only when CRD deletion is successful. Is that okay?

Reply:
@lintongj The purpose of deleting the volume and CRD here is to cleanup whatever we did till now during the CreateVolume call. It was Sandeep's suggestion to consider CreateVolume like a transaction - if there is an error, rollback what ever we have done till now.